localstack/services/apigateway/helpers.py

```diff
@@ -706,25 +708,34 @@ def get_resource_for_path(path: str, path_map: Dict[str, Dict]) -> Optional[Tupl
 
     # if there are no matches, it's not worth to proceed, bail here!
     if not matches:
-        return None
+        return None, None
 
     # so we have matches and perhaps more than one, e.g
     # /{proxy+} and /api/{proxy+} for inputs like /api/foo/bar
     # /foo/{param1}/baz and /foo/{param1}/{param2} for inputs like /for/bar/baz
     if len(matches) > 1:
-        # check if we have an exact match (exact matches take precedence)
-        for match in matches:
-            if match[0] == path:
-                return match
-
-        # not an exact match but parameters can fit in
+        filtered_matches = []
         for match in matches:
-            if path_matches_pattern(path, match[0]):
-                return match
-
-        # at this stage, we have more than one match but we have an eager example like
-        # /{proxy+} or /api/{proxy+}, so we pick the best match by sorting by length
-        sorted_matches = sorted(matches, key=lambda x: len(x[0]), reverse=True)
+            match_methods = list(match[1].get("resourceMethods", {}).keys())
+            # only look for path matches if the request method is in the resource
+            if method in match_methods or "ANY" in match_methods:
+                # check if we have an exact match (exact matches take precedence) if the method is the same
+                if match[0] == path:
```

Comment:
nit: given we are not doing significant work on the different switches it can be written like this
```
if (match[0] == path or match[0] != path and path_matches_pattern(path, match[0])):
    return match
```